### DISCUSSION TITLE
Understand how openssl uses AES sbox

### DISCUSSION THREAD

**Discussion by vm2p** [openssl/openssl#25029] [Q&A] [answered]
Hello,

I some questions related to openssl building/compilation and how openssl makes use of the AES sbox.

Bellow is a (very simple) C program calling AES CTR 128 using EVP:

```
#include <openssl/conf.h>
#include <openssl/evp.h>
#include <openssl/err.h>
#include <string.h>

void handleErrors(void)
{
    ERR_print_errors_fp(stderr);
    abort();
}

int encrypt(unsigned char *plaintext, int plaintext_len, unsigned char *key,
            unsigned char *iv, unsigned char *ciphertext)
{
    EVP_CIPHER_CTX *ctx;
    int len;
    int ciphertext_len;

    if(!(ctx = EVP_CIPHER_CTX_new()))
        handleErrors();

    if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_128_ctr(), NULL, key, iv))
        handleErrors();

    if(1 != EVP_EncryptUpdate(ctx, ciphertext, &len, plaintext, plaintext_len))
        handleErrors();
    ciphertext_len = len;

    if(1 != EVP_EncryptFinal_ex(ctx, ciphertext + len, &len))
        handleErrors();
    ciphertext_len += len;

    EVP_CIPHER_CTX_free(ctx);

    return ciphertext_len;
}

int main (void)
{
    unsigned char key[]={0x00,0x11,0x22,0x33,0x44,0x55,0x66,0x77,0x88,0x99,0xAA,0xBB,0xCC,0xDD,0xEE,0xFF}; 
    unsigned char iv[]={0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00}; 
    unsigned char *plaintext = (unsigned char *)"hello, this is patrick";
    unsigned char ciphertext[128];

    ciphertext_len = encrypt (plaintext, strlen ((char *)plaintext), key, iv, ciphertext);

    printf("Ciphertext is:\n");
    for (int i = 0; i < ciphertext_len; i++) {
        printf("%02x", ciphertext[i]);
    }
    printf("\n");

    return 0;
}
```

I'm compiling the program with the following command:
```
gcc -o prog prog.c -static -no-pie -L/path/to/static/libcrypto -I/path/to/include -l:libcrypto.a
```

However, when analyzing the binary, I can't find the AES sbox anywhere. Concretely, I thought that Te0, Te1, Te2, etc from the [aes_core.c](https://github.com/openssl/openssl/blob/master/crypto/aes/aes_core.c) file but those matrices are not even loaded to the binary, even though I'm statically linking libcrypto.

So, my question is, how does AES sbox is loaded to the binary? Is there a way to force openssl to use Te0, Te1, Te2, etc? I'm configuring openssl with the no-asm --debug options.

Thank you very much for your help.

**Comment by nhorman** [comment, marked as answer]
A few notes:
1) ciphertext_len isn't defined in the main routine
2) The compliation command doesn't work ( -l:libcrypto.a isn't the right syntax)

if I fix up (1) and build with the following command:
```
gcc -o prog prog.c  -no-pie -I./include -L. libcrypto.a
```
Note: I needed to omit -static, as I don't have a static libc available on my workstation

But regardless, you can see libcrypto gets statically linked in:
```
nhorman@hmsbeagle:~/git/openssl$ ldd prog
	linux-vdso.so.1 (0x00007f995c216000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f995c002000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f995c218000)
```
and the sbox symbols are present, exactly from where you thought they would be:
```
nhorman@hmsbeagle:~/git/openssl$ objdump -t prog | grep Te0
00000000006830c0 l     O .rodata	0000000000000400              Te0
```

How are you searching for these symbols?

**Reply by vm2p** [reply]
Hi,

Thank you very much for your answer!

I do see the Te0 values, but they seem to never be used by the `encrypt` function. 

I wrote a series of functions using `angr` to perform a hierarchy search for the functions that call Te0 (and then the functions that call the functions that call `Te0`, and so on and so forth) and there is no path starting on the `encrypt` function that reach Te0.

Instead, it seems to be using a series of functions whose name starts with `__func__.X`, where X is a decimal value. Concretely, it seems to be using `__func__.9640` and `__func__.7909` and not the sbox values from `Te0`.

**Reply by nhorman** [reply]
Te0 isn't a function, its an array, and you can see its use in AES_encrypt and AES_set_encrypt_key.

I'm not  sure how you're using angr, but those __func__ symbols aren't text functions, they exist in the .rodata section of the elf file, and represent stringified names of functions.

If you run:
```
objdump -t <your application>
```
and search for a __func__ symbol, you'll see the offset associated with it.  If you then dump the .rodata section:
```
objdump -s -j .rodata <your application>
```
And look up that offset, you'll find the name of the function it represents in the string table

For instance, if I dump the symbol table of libcrypto.so.3, I find this:
```
000000000040df98 l     O .rodata    000000000000000c              __func__.3
```
Which is a local symbol in a compilation unit with an offset value of 40df98

If I then dump the rodata section and lookup that offset:
```
40df90 00000000 00000000 41534e31 5f693264  ........ASN1_i2d
 40dfa0 5f667000 00000000 00000000 00000000  _fp
```

__func__.3 represents the stringified name of the function ASN1_i2d_fp
